### DISCUSSION TITLE
How work TriggerEvent in plugin ReCAPTCHA for ajax call form?

### DISCUSSION THREAD

**Discussion by korenevskiy** [joomla/joomla-cms#31868] [Q&A] [answered]
Created an ajax module. After receiving the ajax response, an error was received by the **`onCheckAnswer`** method execution plugin.
In the site configuration settings, reCAPTCHA is configured recaptcha_invisible.
So **`$plugin`** is the plugin object **`PlgCaptchaRecaptcha_Invisible`**
But when executing the **`JFactory::getApplication()->triggerEvent('onCheckAnswer');`**
An error occurs in the **`onCheckAnswer`** method of the **`PlgCaptchaRecaptcha`** class
Why is **`Recaptcha_Invisible`** in the site configuration, and the trigger calls a method of the **`Recaptcha`** class???

```
        use Joomla\CMS\Factory as JFactory;
        use Joomla\Registry\Registry as JRegistry;
        use \Joomla\CMS\Plugin\PluginHelper as JPluginHelper;

        JFactory::getDocument()->setBase(JUri::root());

JPluginHelper::importPlugin('captcha');
$captcha_type = JFactory::getConfig()->get('captcha',false);
        if(empty($captcha_type))
            return FALSE;
$plugin = JPluginHelper::getPlugin('captcha', $captcha_type);
        if(empty($plugin) || empty($plugin->params))
            return FALSE;
$plugin->params = new JRegistry($plugin->params);
$param = $plugin->params->toObject();
        if(!$param->public_key || !$param->private_key)
            return FALSE;
$is_valid = JFactory::getApplication()->triggerEvent('onCheckAnswer');
```

**Comment by Fedik** [comment, marked as answer]
Capthca plugins does not trigger an events, they used as callback.
You have to use `Joomla\CMS\Captcha\Captcha` Class.
[Look for example `CaptchaField` ](https://github.com/joomla/joomla-cms/blob/bd4a56ea40047113a8d8b2ceb2689542f9c03a7e/libraries/src/Form/Field/CaptchaField.php#L139) and `CaptchaRule` for validation.

I do not think it is right place for such questions.

**Reply by korenevskiy** [reply]
This is a control for the admin control panel. Even if you place this control on the production page.
Then still this control does not perform the check. NOT check. Thanks to this control, the following happens::
1. Download the Google Recaptcha JS script,
2. Then the public key is checked on the Google side.
3. Then the Recaptcha script returns a response from Google in the form of a test cache in the browser and is placed in the field with 
**`<input name="g-recaptcha-response" type="hidden" value="long_hash">`**
4. This hash must be checked in the side site, verification must be carried out after sending the form to the site in the code of the module, plugin, component.
 
This field does NOT perform validation.
Even if Google checks the public key and sends a verification hash in the browser, this hash will not be of any use if it is not checked when submitting the form. To check the hash when submitting a form, you need to call the plugin trigger(or call function **onCheckAnswer**).
But you're right, the check actually happens in the **[Rule](https://github.com/joomla/joomla-cms/blob/bd4a56ea40047113a8d8b2ceb2689542f9c03a7e/libraries/src/Form/Rule/CaptchaRule.php#L61)** file.
**`joomla-cms/libraries/src/Form/Rule/CaptchaRule.php`** 
But how to use Rule remains a mystery to me.
I'll try to just use just calling the function **`checkAnswer($value);`** 




```
$captcha = Captcha::getInstance((string) $plugin, array('namespace' => (string) $namespace));
return $captcha->checkAnswer($value);
```

**Reply by korenevskiy** [reply]
Rule files are made exclusively for writing forms in XML configuration for modules, plugins, components.
The bottom line is that Rule accepts XML as a parameter.
```
<field name="captcha" type="captcha" validate="captcha"
	namespace="contact"
	label="COM_CONTACT_CAPTCHA_LABEL"
	description="COM_CONTACT_CAPTCHA_DESC" >
</field>
```
Therefore, your suggestion to use FieldCaptcha is not feasible.
But your suggestion helped me find out how to properly call the plugin validation function. 
**Thank you very much.**

**Reply by Fedik** [reply]
> Therefore, your suggestion to use FieldCaptcha is not feasible.

I am not suggested "to use", I have suggested an **example**

**Reply by korenevskiy** [reply]
I tried to write a ready-made answer here, with examples. If anyone has a problem in the future. That was clear from this question-answer. Thank you
